### DISCUSSION TITLE
Does TCP benefit from fixed writes?

### DISCUSSION THREAD

**Discussion by cmazakas** [axboe/liburing#870] [Q&A] [answered]
I'm working on a small HTTP/TCP server and I was wondering, do fixed writes provide any benefit over the normal ones?

I'm reading the docs and they mention fixed writes are useful in conjunction with O_DIRECT which doesn't apply. Do we still get benefits from using fixed buffers with network sends?

**Comment by axboe** [comment, marked as answer]
No benefit for that right now, registered buffers (and the associated read/write fixed) is purely an optimization regular files with O_DIRECT. It does make a considerable difference there. For buffered writes, it doesn't matter, and for socket it also has no effect.

**Reply by cmazakas** [reply]
Ha, thank you for the fast and informative reply.

I was reading the docs and this was my intuition. Thank you for confirming.

**Reply by redbaron** [reply]
Does it apply to zero copy send to socket? I've read somewhere that ZC operations should use fixed buffers

**Reply by isilence** [reply]
Unlike normal sends / socket write, `IORING_OP_SEND_ZC` benefits from using fixed buffers, i.e. no per IO page pinning and lookups
